### PR TITLE
Share a common class_hierarchyt instance across multiple users

### PR DESCRIPTION
**Pull request by smowton**
This is an alternative to #3204, see that PR for details.

### CODE REVIEW

**Comment by smowton** [comment]
@thk123 note this breaks test-gen; the fix is to mirror the change made in `jbmc_parse_options.cpp` (owning a `class_hierarchyt` instance which is then passed to `remove_exceptions` / `remove_instanceof`).

**Comment by tautschnig** [comment]
Would it be possible to have the first four commits in a separate PR? It seems those should go in without debate and wouldn't break TG either. Only the last one is really what #3204 was about?!

**Comment by smowton** [comment]
@tautschnig they're just incidental cleanups in the immediate vicinity of the main change, which I split into their own commits to avoid dirtying the main functional ones. I don't think they're significant enough to bother making their own PR.

**Comment by tautschnig** [comment]
Up to you, but I think we might become better at getting changes merged if we were to arrive at a one-commit-per-PR model. There's a cost in rebasing changes to PR authors and re-reviewing to reviewers.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: fff81e8bf959b870144c060a2755c9894401f25f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88804165
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by peterschrammel — COMMENTED** [review]
Looks good, but needs a TG bump, @thk123.

File: jbmc/src/java_bytecode/remove_exceptions.cpp [hunk 1]
```diff
@@ -589,6 +607,7 @@ void remove_exceptions(
     };
   remove_exceptionst remove_exceptions(
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_exceptions.cpp:608]: :bulb: you might want to precondition class_hierachy if `type == remove_exceptions_typest::REMOVE_ADDED_INSTANCEOF` so that the caller gets a clearer error message.

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 2]
```diff
@@ -529,6 +529,9 @@ int jbmc_parse_optionst::doit()
       *this, options, get_message_handler());
     lazy_goto_model.initialize(cmdline, options);
 
+    class_hierarchy = util_make_unique<class_hierarchyt>();
+    (*class_hierarchy)(lazy_goto_model.symbol_table);
```
Comment (thk123) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: :confused: Can't think of a better way to write this, but is a bit of a head scratcher. :bulb: Perhaps pull into a static method in `class_hierachy.h`: std::unique_ptr<class_hierachyt> create_hierachy(const symbol_tablet &)`?

**Review by thk123 — COMMENTED** [review]
This is in place of the analysis cache I assume? Will make a TG bump now. Otherwise LGTM.

**Comment by smowton** [comment]
https://github.com/diffblue/test-gen/pull/2407

File: jbmc/src/java_bytecode/remove_exceptions.cpp (hunk 1 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_exceptions.cpp:608]: Kept the check in the constructor where it's not duplicated, but added explanatory text.

File: jbmc/src/jbmc/jbmc_parse_options.cpp (hunk 2 above)
Comment (smowton) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: Added an extra constructor, take another look?

File: src/goto-programs/class_hierarchy.h [hunk 3]
```diff
@@ -57,6 +57,10 @@ class class_hierarchyt
   void operator()(const symbol_tablet &);
 
   class_hierarchyt() = default;
+  explicit class_hierarchyt(const symbol_tablet &symbol_table)
```
Comment (thk123) [inline src/goto-programs/class_hierarchy.h:60]: :green_heart: Why even have the default constructor + operator? This is perfect!

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 4]
```diff
@@ -606,12 +609,13 @@ int jbmc_parse_optionst::get_goto_program(
       *this, options, get_message_handler());
     lazy_goto_model.initialize(cmdline, options);
 
+    class_hierarchy = util_make_unique<class_hierarchyt>();
```
Comment (thk123) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: :pick: Use the new constructor

**Review by thk123 — APPROVED** [review]
Thanks for making the TG bump

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 9998ed81409b8bfdbaec532063871bc3418c647a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88821470
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 734dedffdc48bdc27fca02fb5f74e9824f3d7610).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88824136
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/goto-programs/class_hierarchy.h (hunk 3 above)
Comment (smowton) [reply src/goto-programs/class_hierarchy.h:60]: for cases where the user wants to allocate it in advance, e.g. as a member of another class that won't have access to the symbol table until its own `operator()`, so I won't kill the no-arg constructor. However agreed this is better when possible.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 1b43a8a0ea968f63da5b2a9f73f82f7b727c6638).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88835479
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 5]
```diff
@@ -606,12 +609,13 @@ int jbmc_parse_optionst::get_goto_program(
...
+    class_hierarchy =
+      util_make_unique<class_hierarchyt>(lazy_goto_model.symbol_table);
+
     // Show the class hierarchy
     if(cmdline.isset("show-class-hierarchy"))
     {
-      class_hierarchyt hierarchy;
-      hierarchy(lazy_goto_model.symbol_table);
-      show_class_hierarchy(hierarchy, ui_message_handler);
+      show_class_hierarchy((*class_hierarchy), ui_message_handler);
```
Comment (owen-mc-diffblue) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: Why the extra brackets?

Comment (smowton) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: oops, fixed

**Comment by smowton** [comment]
Merging blocked for now because test-gen can't be validated due to problems with a previous PR merged against cbmc/develop.

**Comment by smowton** [comment]
Rebased on fixed cbmc/develop

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 5e9aba835839ca40ca50d5dc16ac9e0477279178).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88984909
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 6]
```diff
@@ -552,6 +552,9 @@ int jbmc_parse_optionst::doit()
       *this, options, get_message_handler());
     lazy_goto_model.initialize(cmdline, options);
 
+    class_hierarchy =
+      util_make_unique<class_hierarchyt>(lazy_goto_model.symbol_table);
```
Comment (owen-mc-diffblue) [inline jbmc/src/jbmc/jbmc_parse_options.cpp:556]: Why is this needed here?
